### PR TITLE
Fix issues with Multi BMI and nested module initialization

### PR DESCRIPTION
**Pull request by robertbartel**
Fixing a few bugs related to this, in particular related to how nested formulations were being cast to a generic super-type to be held in the collection within _Bmi_Multi_Formulation_.

Also including a multi BMI example realization config.

### CODE REVIEW

**Review by hellkite500 — APPROVED** [review]
I will approve this.  There are some potential clean up commits that could be added, but they can be in subsequent PR's if this one needs to be prioritized.

**Comment by robertbartel** [comment]
I'm perplexed at the moment by the current failures of the checks.  On both my Mac and apd-dev1, I have successfully built the main `ngen` executable, and successfully built and run the `test_unit` and `test_bmi_c` executables, with all the tests passing.  I've also build, run, and passed the Fortran tests on just my Mac.

File: src/realizations/catchment/Bmi_Multi_Formulation.cpp [hunk 1]
```diff
@@ -385,10 +367,32 @@ double Bmi_Multi_Formulation::get_response(time_step_t t_index, time_step_t t_de
...
+            if (i == index) continue;
+            out_var_names = modules[i]->get_output_variable_names();
+            if (std::find(out_var_names.begin(), out_var_names.end(), get_bmi_main_output_var()) != out_var_names.end()) {
+                index = i;
+                break;
+            }
+        }
+    }
+    // With the module index, we can also pick the type
+    if (module_types[index] == "bmi_c") {
```
Comment (hellkite500) [inline src/realizations/catchment/Bmi_Multi_Formulation.cpp:385]: Each of these formulation types is MACRO conditionally included, so these lines need appropriate macro guards for each one.

File: include/realizations/catchment/Formulation_Manager.hpp [hunk 2]
```diff
@@ -346,8 +346,45 @@ namespace realization {
 
                 // Attempt to open the directory for evaluation
                 directory = opendir(path.c_str());
+                // Allow for a few retries in certain failure situations
+                size_t attemptCount = 0;
+                std::string errMsg;
+                while (directory == nullptr && attemptCount++ < 5) {
+                    // For several error codes, we should break immediately and not retry
+                    if (errno == ENOENT) {
```
[the remark this answers is not in the document]
Comment (robertbartel) [reply include/realizations/catchment/Formulation_Manager.hpp:354]: `errno` gets set by [opendir](https://man7.org/linux/man-pages/man3/opendir.3.html) in this case and is made accessible by including `<errno.h>`.  See [here](https://man7.org/linux/man-pages/man3/errno.3.html) for more.

**Pull request merged**